prusti-encoder/src/encoders/type/ty_impure.rs

```diff
@@ -0,0 +1,300 @@
+use prusti_rustc_interface::{
+    middle::ty,
+    abi,
+};
+use task_encoder::{EncodeFullResult, TaskEncoder, TaskEncoderDependencies};
+
+use crate::encoders::{lifted::{casters::{CastTypeImpure, CastTypePure}, rust_ty_cast::{GenericCasterImpure, GenericCasterPure, RustTyCastersEnc}}, predicate::{PredicateEnc, PredicateEncDataEnum, PredicateEncDataImmRef, PredicateEncDataMutRef, PredicateEncDataStruct}, PredicateEncOutput, PredicateEncOutputRef};
+
+use super::{
+    lifted::{
+        generic::LiftedGeneric,
+        ty::{EncodeGenericsAsLifted, LiftedTy, LiftedTyEnc},
+    },
+    most_generic_ty::extract_type_params,
+};
+
+pub struct TyImpureEnc;
+
+#[derive(Clone)]
+pub struct TyImpureEncOutputRef<'vir> {
+    /// The predicate output for the "most generic version" of the input type
+    generic_predicate: PredicateEncOutputRef<'vir>,
+
+    pub indirect_predicate: Option<(
+        vir::ExprGenBool<'vir, vir::ExprRef<'vir>, vir::ExprKind<'vir>>,
+        vir::ExprGenBool<'vir, vir::ExprRef<'vir>, vir::ExprKind<'vir>>,
+    )>,
+
+    /// The lifted representation of the input type, as a Viper value
+    pub ty: LiftedTy<'vir, LiftedGeneric<'vir>>,
+
+    pub f_ty: GenericCasterPure<'vir>,
+
+    params: Vec<GenericCasterImpure<'vir>>,
+}
+
+pub struct TyImpureDataStruct<'a, 'vir> {
+    params: &'a [GenericCasterImpure<'vir>],
+    ty_args: Vec<vir::ExprTyVal<'vir>>,
+
+    ref_to_pred: vir::PredicateIdn<'vir, (vir::Ref, vir::ManyTyVal)>,
+    inner: Option<PredicateEncDataStruct<'vir>>,
+}
+
+pub struct TyImpureDataEnum<'vir> {
+    inner: PredicateEncDataEnum<'vir>,
+    ty_args: Vec<vir::ExprTyVal<'vir>>,
+}
+
+pub struct TyImpureDataImmRef<'vir> {
+    inner: PredicateEncDataImmRef<'vir>,
+    ty_args: Vec<vir::ExprTyVal<'vir>>,
+}
+
+pub struct TyImpureDataMutRef<'vir> {
+    inner: PredicateEncDataMutRef<'vir>,
+    ty_args: Vec<vir::ExprTyVal<'vir>>,
+}
+
+impl<'vir> TyImpureEncOutputRef<'vir> {
+    /// Generates a call to `method_assign`, which asserts that the snapshot of
+    /// `self_ref` is `self_new_snap`. Appropriate type arguments are used.
+    pub fn apply_method_assign<'tcx>(
+        &self,
+        vcx: &'vir vir::VirCtxt<'tcx>,
+        self_ref: vir::ExprRef<'vir>,
+        self_new_snap: vir::ExprSnap<'vir>,
+    ) -> vir::Stmt<'vir> {
+        //assert_eq!(self_ref.ty(), &TypeData::Ref);
```

Comment:
```suggestion
```

No need because this is already enforced by `ExprRef`, right?

Tangentially, I wonder: do we want to finally introduce typed `Ref`s at least in our VIR representations that would check things similarly to the snapshots?